src/inmanta/module.py

```diff
@@ -86,6 +88,48 @@
 ModuleName = NewType("ModuleName", str)
 
 
+TModule = TypeVar("TModule", bound="Module")
+TInmantaModuleRequirement = TypeVar("TInmantaModuleRequirement", bound="InmantaModuleRequirement")
+
+
+class InmantaModuleRequirement:
+    def __init__(self, requirement: Requirement) -> None:
+        if requirement.project_name.startswith(ModuleV2.PKG_NAME_PREFIX):
+            raise ValueError("InmantaModuleRequirement instances work with inmanta module names, not python package names.")
+        self._requirement: Requirement = requirement
+
+    @property
+    def project_name(self) -> str:
+        # Requirement converts all "_" to "-". Inmanta modules use "_"
+        return self._requirement.project_name.replace("-", "_")
+
+    @property
+    def key(self) -> str:
+        # Requirement converts all "_" to "-". Inmanta modules use "_"
+        return self._requirement.key.replace("-", "_")
+
+    @property
+    def specifier(self):
+        return self._requirement.specifier
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, InmantaModuleRequirement):
+            return NotImplemented
+        return self.project_name == other.project_name and self._requirement == other._requirement
+
+    def __contains__(self, version: str) -> bool:
+        return version in self._requirement
+
+    def __str__(self) -> str:
+        return str(self._requirement).replace("-", "_")
+
+    @classmethod
+    def parse(cls: Type[TInmantaModuleRequirement], spec: str) -> TInmantaModuleRequirement:
+        if "-" in spec:
+            raise ValueError("Invalid inmanta module requirement: inmanta module names use '_', not '-'.")
```

Comment:
We can't because `Requirement.parse` transforms all underscores to dashes, so it needs to be checked before constructing the `Requirement` instance, i.e. when parsing. If you think this can be improved I'm open to suggestions.